### PR TITLE
Support `replace` in `strings_udf`

### PR DESCRIPTION
**Pull request by brandon-b-miller**
This PR adds support for the following function in `strings_udf`:

- `str.replace`

Part of https://github.com/rapidsai/cudf/issues/9639

### CODE REVIEW

File: python/cudf/cudf/core/udf/strings_lowering.py [hunk 1]
```diff
@@ -50,6 +51,41 @@ def masked_len_impl(context, builder, sig, args):
...
+
+    src_masked = cgutils.create_struct_proxy(sig.args[0])(
+        context, builder, value=args[0]
+    )
+    to_replace_masked = cgutils.create_struct_proxy(sig.args[0])(
+        context, builder, value=args[1]
+    )
+    replacement_masked = cgutils.create_struct_proxy(sig.args[0])(
+        context, builder, value=args[2]
+    )
```
Comment (shwina) [inline python/cudf/cudf/core/udf/strings_lowering.py:71]: Some of the boilerplatey-ness here could be reduced with something like:

```python
src_masked, to_replace_masked, replacement_masked = create_struct_proxies(context, builder, sig.args[0], args)
```

Comment (brandon-b-miller) [reply python/cudf/cudf/core/udf/strings_lowering.py]: something like https://github.com/rapidsai/cudf/pull/12207/commits/51ba4e3b738c0939ea723bebf4fbc13a22fa9790 ?

**Review by shwina — APPROVED** [review]

**Comment by brandon-b-miller** [comment]
rerun tests

File: python/strings_udf/cpp/src/strings/udf/shim.cu [hunk 2]
```diff
@@ -329,3 +330,20 @@ extern "C" __device__ int concat(int* nb_retval, void* udf_str, void* const* lhs
...
+  auto src_ptr         = reinterpret_cast<cudf::string_view const*>(src);
+  auto to_replace_ptr  = reinterpret_cast<cudf::string_view const*>(to_replace);
+  auto replacement_ptr = reinterpret_cast<cudf::string_view const*>(replacement);
+
+  auto udf_str_ptr = new (udf_str) udf_string;
+  udf_string result;
+
+  result = replace(*src_ptr, *to_replace_ptr, *replacement_ptr);
+
+  *udf_str_ptr = result;
```
Comment (davidwendt) [inline python/strings_udf/cpp/src/strings/udf/shim.cu:346]: The compiler may optimize this for us but changing to this _may_ prevent an extra alloc/copy:
```suggestion
  *udf_str_ptr = replace(*src_ptr, *to_replace_ptr, *replacement_ptr);
```

**Comment by brandon-b-miller** [comment]
@gpucibot merge

**Pull request merged**